File: tasks/builtin/bug-fix-03-cs/Starter.cs
using System.Threading;

public class ThreadSafeCounter
{
    private int _value = 0;
    private readonly object _lock = new object();

    public void Increment()
    {
        int current = _value;
        Thread.Sleep(1);
        _value = current + 1;
    }

    public void Decrement()
    {
        lock (_lock)
        {
            _value--;
        }
    }

    public int GetValue()
    {
        return _value;
    }

    public void Reset()
    {
        lock (_lock)
        {
            _value = 0;
        }
    }

    public void IncrementBy(int n)
    {
        int current = GetValue();
        lock (_lock)
        {
            _value = current + n;
        }
    }
}

File: tasks/builtin/bug-fix-03-js/starter.js
class AsyncCounter {
    constructor() {
        this.value = 0;
    }

    async increment() {
        const current = this.value;
        await new Promise(resolve => setTimeout(resolve, 0));
        this.value = current + 1;
    }

    async decrement() {
        this.value -= 1;
    }

    get() {
        return this.value;
    }

    reset() {
        this.value = 0;
    }

    async incrementBy(n) {
        const current = this.get();
        await new Promise(resolve => setTimeout(resolve, 0));
        this.value = current + n;
    }
}

module.exports = { AsyncCounter };
